### DISCUSSION TITLE
Separate data models from domain models

### DISCUSSION THREAD

**Discussion by abdu292** [jasontaylordev/CleanArchitecture#674] [Q&A]
It's just a question. I used this template for two projects so far and they were all fine. However on an another project that includes a lot of domain logics I would want to separate ef persistence data models from business models. This would give me the complete flexibility to bend my domain models however I want without leaving any trace of persistence knowledge in the domain objects.
Is this possible without introducing a new repository** layer && using this template? 

If it is doable, any direction as to how to do it cleanly will be very much appreciated. Thanks a lot.

**I understand I can do this using separate repos such as IReadRepo & IWriteRepo then converting the domain models to persisted models & vice-versa. However, I'd like to stick with DbSet & EF provided repos. I'll be using EF for the lifetime of this project.

**Comment by neman** [comment]
> This would give me the complete flexibility to bend my domain models however I want without leaving any trace of persistence knowledge in the domain objects.
> 

Can you please explain what do you mean by `trace of persistence knowledge in the domain objects`? Domain model does not know anything about persistence. Persistence configuration is in infrastructure layer, it references domain and bend it as you wish.

**Comment by abdu292** [comment]
@neman Sure I can explain with a simple example.

I have a domain **Student** with different logics in it as to which type of student can enroll.
Here is my persistence entity model (which is a very simple model).

```
public class Student
{
    public int Id { get; set; }
   public string Name { get; set; }
   public EnrolledSemesterId { get;set; }
}
```

However my domain model consists of three classes (note the first one is an abstract class)

```
public abstract Student
{
  public string Name { get; set; }
  public Semester EnrolledSemester { get; set; }

  public abstract bool CanEnroll(Semester semester);
}

public class DomesticStudent : Student
{
    public override bool CanEnroll(Semester semester) { // specific logic  }
}

public class InternationalStudent : Student
{
    public override bool CanEnroll(Semester semester) { //specific logic  }
}
```

Here as you can see I want my domain model to perform its operations the way I outlined above. However my persistence is just a simple class with three fields. Let me know if you need more information. 

Thanks for your time.

**Reply by rmunson8** [reply]
Hello, we are starting a new product using Clean Architecture and have exactly this question, as well.  I have read several discussion threads online where the commenter makes an assumption that the models used by EF (or any infra layer) should just move to the application layer, but @abdu292 shows a great example where the domain and infra models are different. Why wouldn't the default be that the interface that gets created in the Application layer for use by the Infrastructure layer to implement also needs to have a method signature that uses domain models and it's the responsibility of the infra layer to map to/from infra models.

Thoughts @jasontaylordev?

Thank you!
